peer.go

```diff
@@ -345,6 +346,14 @@ func (p *Peer) GetConnection(ctx context.Context) (*Connection, error) {
 		return activeConn, nil
 	}
 
+	p.newConnLock.Lock()
```

Comment:
Can you add a comment above this saying why we want to do this (to restrict pending connection attempts per peer)